[receiver/hostmetrics] add process.handles metric

braydonk opened the pull request:
**Description:** <Describe what has changed.>
Adds a new Windows-exclusive metric called process.handles, which represents the handle count of the given process. When enabled, the receiver will make a WMI Query at the beginning of each scrape to update the handle count for all processes on the system. If the metric is enabled on a platform other than Windows, an error will be produced when attempting to refresh handle counts. This matches the rough behaviour of the Linux exclusive `process.open_file_descriptors`.

**Link to tracking Issue:** <Issue number if applicable>
#21379 
#12482

**Testing:** <Describe what testing was performed and which tests were added.>
Ran the binary with the following configuration:
```
receivers:
  hostmetrics:
    collection_interval: 2s
    scrapers:
      cpu: {}
      disk: {}
      filesystem: {}
      load: {}
      memory: {}
      network: {}
      paging: {}
      process:
        mute_process_name_error: true
        metrics:
          process.handles:
            enabled: true
      processes: {}

exporters:
  file:
    path: x.json

service:
  pipelines:
    metrics:
      receivers: [hostmetrics]
      exporters: [file]
```

The following is an example result of a scrape with this configuration.
https://gist.github.com/braydonk/c97996272574319e03111dc79076a1bd

**Documentation:** <Describe the documentation added.>
`documentation.md` was generated by `mdatagen` from the `metadata.yaml` updates.

braydonk committed:
hostmetricsreceiver: add process.handles metric

Adds a new Windows-exclusive metric called process.handles, which
represents the handle count of the given process. When enabled, the
receiver will make an additional syscall, specifically
`NtQuerySystemInformation` at the beginning of each scrape
to update the handle count for all processes on the system. If the
metric is enabled on a platform other than Windows, an error will be
produced when attempting to refresh handle counts. This matches the
rough behaviour of the Linux exclusive `process.open_file_descriptors`.

braydonk commented:
Not sure about that lint failure. I don't think it really matters that an `err` declaration is shadowed... Either way, that failure doesn't appear to be cause by this PR but I can fix it if needed.

jefferbrecht commented on receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:
```diff
@@ -233,6 +242,10 @@ func (s *scraper) getProcessMetadata() ([]*processMetadata, error) {
 		data = append(data, md)
 	}
 
+  if err := s.refreshHandleCounts(); err != nil {
```
```suggestion
	if err := s.refreshHandleCounts(); err != nil {
```

jefferbrecht replied:
Also, wouldn't the first `scrape` always be empty since `refreshHandleCount` is called after `scrapeAndAppendHandlesMetric`?

braydonk committed:
processscraper: handle count refresh in right spot

I put the handle count refresh after all the metrics scrapes instead of
after collecting all process handles, which meant the first scrape would
always be empty.

braydonk committed:
handles: use WMI to get handle counts instead

NtQuerySystemInformation is an unsupported internal API. This means it
can be subject to change. This commit swaps handle count retrieval to
use Windows Management Instrumentation to get the counts instead.

dmitryax commented on receiver/hostmetricsreceiver/internal/scraper/processscraper/metadata.yaml:
```diff
@@ -164,6 +164,14 @@ metrics:
+    description: Number of handles held by the process.
+    extended_documentation: This metric is only available on Windows.
+    unit: '{count}'
+    gauge:
```
I believe this should be non-monotonic sum

braydonk replied:
Most Windows users who need to know their handle count will check it in something like ProcExp or the performance counters and see the handle count at the current point in time. I don't think I understand why non-monotonic sum would be used here over gauge for that.

braydonk replied:
I think I understand non-monotonic sums more now. I can make that change to conform with `process.open_file_descriptors`.

braydonk replied:
Done in 44c7bab

dmitryax commented:
Please take a look at the failing CI

braydonk committed:
handles: introduce handleCountQueryer interface

Introduces a more generic interface for handle count querying than what
was there before. This makes it easier to reason about when using an
implementation with different syscalls, so the handleCountManager
implementation can remain the way it was. Fixed the test to reflect the
changes.

braydonk committed:
handles: fix unused pid arg in unsupportedManager

The argument needs to be there to implement the interface but it doesn't
need to be named because it's unused.

braydonk committed:
handles: run goporto

braydonk committed:
hostmetricsreceiver: change process.handles to sum

Changing the process.handles metric from a gauge to a cumulative
non-monotonic sum.

braydonk committed:
run make gotidy

jefferbrecht commented on receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:
```diff
@@ -167,6 +176,10 @@ func (s *scraper) getProcessMetadata() ([]*processMetadata, error) {
 
 	var errs scrapererror.ScrapeErrors
 
+	if err := s.refreshHandleCounts(); err != nil {
```
I might be misunderstanding what `getProcessHandles` does, but isn't this refresh still too late? It seems like `handles` would always be empty on the first invocation of `getProcessMetadata`.

jefferbrecht replied:
Ah thanks, I got mixed up between handles and handle counts there.

braydonk committed:
handlecount: rename handles package

The `handles` package name actually clashed with a ton of other places
that had variables called `handles`. This commit renames the package to
`handlecount`.

dmitryax reviewed (approved):
LGTM. Can you also please submit an issue for the new metric name in semantic conventions and link it to https://github.com/open-telemetry/opentelemetry-specification/issues/3556?

dmitryax commented:
Please rebase

braydonk commented:
Thanks for the review @dmitryax! I've rebased and will wait to ensure CI checks pass again.

I opened a semantic conventions PR and linked it to the working group issue. I've heard that these metrics are being moved to being generated from YAML instead of written directly in the markdown document, so I may delay that PR until that work is done and move it to a YAML definition.

braydonk commented:
Rebased again to pick up the k8sclusterreceiver metadata fix.

braydonk commented:
@dmitryax Looks like with the metadata fix from main, CI is green now. I am unable to merge, hopefully it can get merged before I need to potentially rebase again.
